design-documents/checkout/alternative-checkout-flow.md

```diff
@@ -0,0 +1,50 @@
+# Alternative checkout flow
+
+In scope of work on GraphQL and storefront APIs we have an opportunity to improve design and features of storefront checkout.
+
+The purpose of this document is to discuss possible alternatives to current Magento checkout flow. 
+
+  1. When Quote is created?
+     * For physical products on Review & Payments step
+     * For virtual products billing address has to be entered first
+  1. Cart properties:
+     * Line Items:
+       * SKU
+       * Selected options (custom/configurable)
+       * Quantity
+       * Regular price
+       * Price
+  4. Applicable cart rules calculator arguments
+     * Cart
+     * Addresses
+     * Date & Time (admin preview)
+  2. Quote factory arguments: Quote
+     * !Line Items
+     * Dimensions (weight & size) (based on LineItems)
+     * Shipping (requited for physical products)
+       * Addresses
+       * Selected Shipping Methods
+     * Billing address (required for virtual products). How to deal with multi-payment?
+     * Coupons
+     * Gift cards
+     * Store credit
+     * Cart rules (calculated by Applicable catalog rules calculator)
+     * Customer
+  2. Totals calculator: Totals
+     * QuoteArgumentDTO
+     * PreviousTotals
+  5. Quote
+     * LineItems
+       * Totals
+       * Adjustments
+           * Taxes
+           * Cart rule discounts
+     * Billing addresses
+     * Shipping addresses
+     * Coupons
+     * Gift cards
+     * Store credit
+     * Customer (optional)
+  6. Place Order: Order
```

Comment:
we allow creating multiple quotas based on the single cart (quote segregation based on addresses), from my standpoint it makes sense to make order placement as an atomic operation which allows submitting multiple quotes. It runs us in another question - multiple payments handling together. but this also has a sense. Orders with different destination mean addresses may have different process flow which is fine. For example, the multi-tenant installation will require such functionality.

Reply:
I second this, quotes split by shipping address should still be payable with a single payment during checkout. This has been a major limitation of the current design and is one of the reasons so few merchants utilize the multi address checkout. Having to do separate payments for each order makes the process incompatible with most current payment processors.